test.py

```diff
@@ -626,68 +624,64 @@ def _reset(self) -> None:
     async def run(self, options: argparse.Namespace) -> Test:
         self.success = False
         self.summary = "failed"
+        self.is_equal_result = False
 
         def set_summary(summary):
             self.summary = summary
             logging.info("Test %s %s", self.uname, summary)
             if self.server_log is not None:
                 logging.info("Server log:\n%s", self.server_log)
 
-        async with self.suite.clusters.instance() as cluster:
-            logging.info("Leasing Scylla cluster %s for test %s", cluster, self.uname)
-            self.args.insert(1, "--host={}".format(cluster[0].host))
-            # If pre-check fails, e.g. because Scylla failed to start
-            # or crashed between two tests, fail entire test.py
-            try:
-                cluster.before_test(self.uname)
-                self.is_before_test_ok = True
-                cluster[0].take_log_savepoint()
-                self.is_executed_ok = await run_test(self, options, env=self.env)
-                cluster.after_test(self.uname)
-                self.is_after_test_ok = True
-
-                if self.is_executed_ok is False:
-                    set_summary("""returned non-zero return status.\n
+        # If pre-check fails, e.g. because Scylla failed to start
+        # or crashed between two tests, fail entire test.py
+        try:
+            await self.harness.start()
+            self.args.insert(1, f"--harness-api={self.harness.sock_path}")
+            self.is_executed_ok = await run_test(self, options, env=self.env)
+            await self.harness.stop()
```

Comment:
stop() is not executed if run_test throws. The cluster itself is protected by an artifact, but it still doesn't look safe, e.g. if we retry a flaky test, we will leak clusters in the process. I think an equivalent change is to replace async with clusters.instance() with async iwth harness.instance(). That would preserve the alignment of the code, simplifying the review. What do you think?

Reply:
Can't it just be called in `finally:` ?

Reply:
Calling in `finally:`